x-pack/plugins/lens/public/indexpattern_datasource/operations/layer_helpers.ts

```diff
@@ -0,0 +1,353 @@
+    field &&
+    ('sourceField' in previousColumn ? field.name !== previousColumn.sourceField : true)
+  ) {
+    const newColumn = operationDefinition.buildColumn({ ...baseOptions, field, previousColumn });
+
+    // TODO: This is probably wrong, not all params are shared on all ops?
+    if ('params' in previousColumn) {
+      newColumn.params = previousColumn.params;
+    }
+
+    if (previousColumn.customLabel) {
+      newColumn.customLabel = true;
+      newColumn.label = previousColumn.label;
+    }
+
+    const newColumns = { ...layer.columns, [columnId]: newColumn };
+    return {
+      ...layer,
+      columnOrder: getColumnOrder({ ...layer, columns: newColumns }),
+      columns: adjustColumnReferencesForChangedColumn(newColumns, columnId),
+    };
+  } else {
+    throw new Error('nothing changed');
+  }
+}
+
+function addBucket(
+  layer: IndexPatternLayer,
+  column: IndexPatternColumn,
+  addedColumnId: string
+): IndexPatternLayer {
+  const [buckets, metrics] = separateBucketColumns(layer);
+
+  if (buckets.length === 0 && column.operationType === 'terms') {
+    column.params.size = 5;
+  }
+
+  const oldDateHistogramIndex = layer.columnOrder.findIndex(
+    (columnId) => layer.columns[columnId].operationType === 'date_histogram'
+  );
+  const oldDateHistogramId =
+    oldDateHistogramIndex > -1 ? layer.columnOrder[oldDateHistogramIndex] : null;
+
+  let updatedColumnOrder: string[] = [];
+  if (oldDateHistogramId && column.operationType === 'terms') {
+    // Insert the new terms bucket above the first date histogram
+    updatedColumnOrder = [
+      ...buckets.slice(0, oldDateHistogramIndex),
+      addedColumnId,
+      ...buckets.slice(oldDateHistogramIndex, buckets.length),
+      ...metrics,
+    ];
+  } else {
+    // Insert the new bucket after existing buckets. Users will see the same data
+    // they already had, with an extra level of detail.
+    updatedColumnOrder = [...buckets, addedColumnId, ...metrics];
+  }
+  return {
+    ...layer,
+    columns: { ...layer.columns, [addedColumnId]: column },
+    columnOrder: updatedColumnOrder,
+  };
+}
+
+function addMetric(
+  layer: IndexPatternLayer,
+  column: IndexPatternColumn,
+  addedColumnId: string
+): IndexPatternLayer {
+  return {
+    ...layer,
+    columns: {
+      ...layer.columns,
+      [addedColumnId]: column,
+    },
+    columnOrder: [...layer.columnOrder, addedColumnId],
+  };
+}
+
+function separateBucketColumns(layer: IndexPatternLayer) {
+  return partition(layer.columnOrder, (columnId) => layer.columns[columnId]?.isBucketed);
+}
+
+export function getMetricOperationTypes(field: IndexPatternField) {
+  return operationDefinitions.sort(getSortScoreByPriority).filter((definition) => {
+    if (definition.input !== 'field') return;
+    const metadata = definition.getPossibleOperationForField(field);
+    if (!metadata) return;
+    return (
+      !metadata.isBucketed && (metadata.dataType === 'number' || metadata.dataType === 'document')
```

Comment:
nit: How can `dataType` be `document`? Seems like this condition can be removed

Reply:
I wrote this because our type definition suggests that this can be set as a metric operation type, but it seems like we have two types that I mixed up here: 1. Types of metrics that we have (all numbers) 2. Types of fields in the index pattern